src/cargo/ops/registry/publish.rs

```diff
@@ -485,6 +586,74 @@ fn transmit(
     Ok(())
 }
 
+// State for tracking dependencies during upload.
+struct PublishOrder {
+    // The reverse-dependency graph. It has an edge p -> q if package q depends on package p.
+    rev_graph: Graph<PackageId, ()>,
+    // The weight of a package is the number of unpublished dependencies it has.
+    weights: HashMap<PackageId, usize>,
+}
+
+impl PublishOrder {
+    // Given a package dependency graph, creates a `PublishOrder` for tracking state.
+    fn new(graph: &Graph<PackageId, ()>) -> Self {
+        let rev_graph = graph.reversed();
+
+        let weights: HashMap<_, _> = rev_graph
+            .iter()
+            .map(|id| (*id, graph.edges(id).count()))
+            .collect();
+        Self { rev_graph, weights }
+    }
+
+    // The set of packages that are ready for publishing (i.e. have no outstanding dependencies).
+    fn ready(&self) -> Vec<PackageId> {
+        let mut ready: Vec<_> = self
+            .weights
+            .iter()
+            .filter_map(|(id, weight)| (*weight == 0).then_some(*id))
```

Comment:
There a reason we do this rather than remove them in `mark_published`?

This is sort of a continuation of https://github.com/rust-lang/cargo/pull/14433#pullrequestreview-2273538538

Reply:
Graph nodes aren't removable (and with the current graph representation, making them removable has possible footguns: either we allow dangling edges to nodes that no longer exist, or we need to scan the whole graph on each node removal, making it easy to accidentally introduce quadratic behavior).

Rather than mess with the graph type, I've added some additional state to `PublishOrder` that hopefully makes things more clear.

Reply:
Ah, thanks for the clarification. Yes, I think this is nicer and the names look good to me. My only concern is that I think this makes the whole process quadratic in the number of packages, since `take_ready` needs to scan over all the remaining packages. There's an intermediate solution that avoids the quadratic behavior by keeping the `ready` field (but still removing the `remaining` field). But also maybe workspaces are small enough to not worry about it...

Reply:
Large workspaces are likely a hundred members or so.  Publish performance isn't in a critical path as well.

Reply:
Sounds good. I've pulled in your changes and squashed them in.